core/src/main/java/org/jdbi/v3/core/SqlArrayArgumentFactory.java

```diff
@@ -23,19 +23,31 @@
 
 class SqlArrayArgumentFactory implements ArgumentFactory {
     @Override
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({ "unchecked", "rawtypes" })
     public Optional<Argument> build(Type type, Object value, StatementContext ctx) {
         Class<?> erasedType = GenericTypes.getErasedType(type);
         if (erasedType.isArray()) {
             Class<?> elementType = erasedType.getComponentType();
             return ctx.findArrayTypeFor(elementType)
-                    .map(arrayType -> new SqlArrayArgument(arrayType, value));
+                    .map(arrayType -> new SqlArrayArgument<>(arrayType, value));
         }
-        if (List.class.isAssignableFrom(erasedType)) {
-            return GenericTypes.findGenericParameter(type, List.class)
+        if (Collection.class.isAssignableFrom(erasedType)) {
+            return guessElementType(type, (Collection<?>) value)
                     .flatMap(ctx::findArrayTypeFor)
-                    .map(arrayType -> new SqlArrayArgument(arrayType, (List) value));
+                    .map(arrayType -> new SqlArrayArgument<>(arrayType, (Collection) value));
         }
         return Optional.empty();
     }
+
+    private Optional<? extends Type> guessElementType(Type type, Collection<?> value) {
```

Comment:
If the element type was a `Class<T>` then users could bind the array without having to use `GenericType`:

```java
statement.bindArray(int.class, new ArrayList<>(1, 2, 3));
```

On principle, I'm a little leery of going by the first element. I think if we can't reliably figure out what the type is, then the user should have to tell us.